plugins/states/generational_state_fact.py

```diff
@@ -0,0 +1,290 @@
+    The dict we receive as argument is then the state dict, that we can
+    assign to the state attribute.
+    """
+
+    state: dict
+
+    def get_state(self) -> dict:
+        return self.state
+
+    @classmethod
+    def build_from_state(cls, state: dict) -> "LegacyStateFact":
+        return cls(state=state)
+
+
+class GenerationalStateFact(StateFact):
+    @classmethod
+    @abc.abstractmethod
+    def generation(cls) -> str:
+        """
+        Should be implemented by the subclass, and return the generation identifier
+        """
+
+    def _iter(self, **kwargs) -> "pydantic.TupleGenerator":
+        """
+        We overwrite the _iter method simply to add our generation marker to the
+        generated dict or json payload.
+        """
+        for x in super()._iter(**kwargs):
+            yield x
+
+        yield STATE_DICT_GENERATION_MARKER, self.generation()
+
+    @classmethod
+    def build_from_state(cls: typing.Type[GSF], state: dict) -> GSF:
+        state_generation = state[STATE_DICT_GENERATION_MARKER]
+        if not state_generation == cls.generation():
+            # Actively check that the state dict is of the correct generation
+            raise ValueError(
+                f"Unexpected generation value: {state_generation} != {cls.generation()}"
+            )
+
+        return cls(**state)
+
+
+class AlbatrossGenerationStateFact(GenerationalStateFact):
+    """
+    This is the first iteration on a generational approach for state dicts.
+    The generation is named "Albatross" because it is an animal starting with
+    letter A.  It is expected for the following generation to respect this
+    scheme and pick another animal with the next letter coming in the alphabetic
+    order.
+    """
+
+    state: dict
+    created_at: datetime.datetime
+    updated_at: datetime.datetime
+    config_hash: str
+
+    def get_state(self) -> dict:
+        return self.state
+
+    @classmethod
+    def generation(cls) -> str:
+        return "Albatross"
+
+
+state_fact_generations: typing.Dict[typing.Optional[str], typing.Type[StateFact]] = {
+    None: LegacyStateFact,
+    AlbatrossGenerationStateFact.generation(): AlbatrossGenerationStateFact,
+}
+"""
+This dict holds all the generations of state dicts which were supported by the module.
+It can be navigated to find which class should be used to load the state stored into
+a python object.
+"""
+
+
+def build_state_fact(raw_state: dict) -> StateFact:
+    """
+    This function can be used to reconstruct a state fact object from its
+    corresponding raw state dictionary..
+    """
+
+    # Get the generation attribute from the raw_state, if it is set
+    generation = raw_state.get(STATE_DICT_GENERATION_MARKER)
+
+    # Get the state fact class corresponding to this generation
+    state_fact_class = state_fact_generations[generation]
+
+    # Build the state fact and return it
+    return state_fact_class.build_from_state(raw_state)
+
+
+def state_converter(
```

Comment:
Why have this fairly complex plumbing, if it could just be on the classes?

Reply:
Using simple functions felt a bit more flexible (you could then have multiple implementations to target a state for example), but we don't need it.  I changed it to a class method in https://github.com/inmanta/terraform/pull/49/commits/7a5af1ad70a2ee46c7a72a493670591122f588ee